run-cgprna/run_cgprna/fusion_tools.py

```diff
@@ -0,0 +1,173 @@
+import os
+import sys
+import re
+import shutil
+import copy
+from string import Template
+from . import run_templates_in_shell, untar, mkdir
+import gzip
+
+BAM_TO_FASTQ = Template('bamtofastq exclude=SECONDARY,SUPPLEMENTARY T=$bam2fq_tmp S=$bam2fq_tmp_single_end O=$bam2fq_tmp_unmatched O2=$bam2fq_tmp_unmatched2 gz=1 level=1 F=$bam2fq_tmp_matched F2=$bam2fq_tmp_matched_2 filename=$in_bam')
+# Defuse and its wapper defuse_fusion.pl cannot handle gzipped file.
+BAM_TO_FASTQ_FOR_DEFUSE = Template('bamtofastq exclude=SECONDARY,SUPPLEMENTARY T=$bam2fq_tmp S=$bam2fq_tmp_single_end O=$bam2fq_tmp_unmatched O2=$bam2fq_tmp_unmatched2 F=$bam2fq_tmp_matched F2=$bam2fq_tmp_matched_2 filename=$in_bam')
+TOPHAT_FUSION = Template('tophat_fusion.pl -s $sample_name -o $out_dir -t $threads -r $reference_data_root -sp $species -rb $ref_build -gb $gene_build $input')
+STAR_FUSION = Template('star_fusion.pl -s $sample_name -o $out_dir -t $threads -r $reference_data_root -sp $species -rb $ref_build -gb $gene_build $input')
+DEFUSE_FUSION = Template('defuse_fusion.pl -s $sample_name -o $out_dir -t $threads -r $reference_data_root -sp $species -rb $ref_build -gb $gene_build $input')
+DEFUSE_FILTER = Template('defuse_filters.pl -s $sample_name -o $out_dir -i $out_dir/${sample_name}.defuse-fusion.normals.filtered.txt')
+
+NORMAL_FUSION_ARG_NAME='-normals'
+
+REF_RELATED_DEFAULTS = {
+    'species': 'unspecified_species',
+    'ref_build': 'unspecified_ref_build',
+    'gene_build': 'ensembl'
+}
+
+
+def validate_input_seq_files(file_names):
+    '''
+    Perl wrappers: tophat_fusion.pl, star_fusion.pl and defuse.pl all support multiple input files but are restricted into a single type, i.e either BAM or FastQ files. As we're converting BAMs to fastqs before passing them to the wappers, we can easily handle a mixture of BAM and FastQ files, but needs to validate the files before starting to convert BAM to FastQ, so we don't waste time and resources on splitting a BAM when some of the FastQ files are not following the Perl wrapper's file name conventions.
+    '''
+    pairs = {}
+    for a_file in file_names:
+        if a_file.endswith('.bam'):
+            continue
+        if not os.path.exists(a_file):
+            sys.exit('Error: can not find input file: %s' % a_file)
+        a_file = os.path.abspath(a_file)
+        match = re.match(r'(.*)_([12])\.f(?:ast)?q(?:\.gz)?$', os.path.basename(a_file))
+        if match.group(2):
```

Comment:
I'd store the matching groups in variables for clarity and performance (since they are used multiple times within this block).

Reply:
You're right, I stupidly thought once I did `re.match`,  `match.group()` is computed and results are stored in memory. Not sure what I was thinking....

Reply:
Actually I believe you're right, it's just that it should be cheaper to call the `group` method once, and the variable name you'd assign it to would help document the identity of that match.